### ISSUE TITLE
[Feature Request]: Batchwise torch.lstsq

### ISSUE THREAD

**Issue by TrentWeiss** [pytorch/pytorch#27749]
## 🚀 Feature
A batchwise verison of torch.lstsq https://pytorch.org/docs/stable/torch.html?highlight=lstsq#torch.lstsq

Exact same interface, but with the option for "input" and "A" to be batches of corresponding least-squares formulations to solve.

## Motivation

I am working on a method of predicting parameterized curves (in particular, Bezier Curves). Fitting a Bezier curve to measured data is pretty easy with some simple least-squares linear algebra, but it would be nice to do it batch-wise. 

## Pitch

torch.lstsq should support batches of Least-Squares problems instead of just a single Least-Squares problem

## Alternatives

I have already implemented the capability I need via torch.svd (which does support batch-wise solving), but it would be nice if PyTorch had this capability built-in, perhaps with a faster/more numerically stable technique under-the-hood

Obviously, something like:
 `
torch.stack([torch.lstsq(B[i], A[i]) for i in range(batch)],dim=0)
`
Would also work, but this seems hacky.

## Additional context

cc @vincentqb @vishwakftw @jianyuh @nikitaved @pearu @mruberry @heitorschueroff @SsnL

**Comment by vishwakftw** [comment]
Hi @Electric-Turtle,

Thank you for opening the issue. This is part of the plan in #7500, and I hope to add this before 1.4 is released.

**Comment by WANGZhaowei-Wesley** [comment]
still not yet

**Comment by jramapuram** [comment]
Any update on this? Seems like a lot of other linalg ops have support, but not the least squares solver: probably one of the most common operands in machine learning.

**Comment by shijianjian** [comment]
Any update?

**Comment by syltruong** [comment]
If you are only looking to compute the `solution` output from [`torch.lstsq`](https://pytorch.org/docs/stable/generated/torch.lstsq.html), but batch-wise, you could make use of [`torch.pinverse`](https://pytorch.org/docs/stable/generated/torch.pinverse.html) which is implemented to support batch-wise computation.

> More about the relationship between the Moore-Penrose pseudo-inverse and ordinary least-squares can be found here: https://en.wikipedia.org/wiki/Moore%E2%80%93Penrose_inverse#Linear_least-squares

So that, if the take the example from the [`torch.lstsq`](https://pytorch.org/docs/stable/generated/torch.lstsq.html) documentation:

# No batch support

```python
import torch

A = torch.tensor([[1., 1, 1],
                  [2, 3, 4],
                  [3, 5, 2],
                  [4, 2, 5],
                  [5, 4, 3]])

B = torch.tensor([[-10., -3],
                  [ 12, 14],
                  [ 14, 12],
                  [ 16, 16],
                  [ 18, 16]])

X, _ = torch.lstsq(B, A)

X[:A.shape[1]]  # According to the doc: "The first n rows of X contains the solution."

# tensor([[2.0000, 1.0000],
#         [1.0000, 1.0000],
#         [1.0000, 2.0000]])
```

# With batch support

```python
def batch_lstsq(
    input: torch.Tensor,  # matrix B of shape (batch * m * k) 
    A: torch.Tensor  # matrix A of shape (batch * m * n) 
):
   
    X = torch.bmm(
        torch.pinverse(A),
        input
    )
 
    return X
```

## Tests

- with the same tensors as the ones defined above, but repeated

```python
n_batch = 3

A_batch = torch.cat(n_batch * [A.unsqueeze(0)])
B_batch = torch.cat(n_batch * [B.unsqueeze(0)])

solution = batch_lstsq(B_batch, A_batch)

for i in range(len(solution)):    
    
    X, _ = torch.lstsq(B_batch[i], A_batch[i])
    X = X[:A_batch[i].shape[1]]
    
    assert torch.allclose(solution[i], X[:A.shape[1]])
```

- with random tensors, works for both (m >= n) and (m < n)

```python
n_batch = 3
m = 10  # 2
n = 3
k = 2

A_batch = torch.rand(n_batch, m, n)
B_batch = torch.rand(n_batch, m, k)

solution = batch_lstsq(B_batch, A_batch)

for i in range(len(solution)):    
    
    X, _ = torch.lstsq(B_batch[i], A_batch[i])
    X = X[:A_batch[i].shape[1]]
    
    assert torch.allclose(solution[i], X[:A.shape[1]])
```

## Wall clock performance test

- Setup
```
n_batch = 1000

A_batch = torch.cat(n_batch * [A.unsqueeze(0)])
B_batch = torch.cat(n_batch * [B.unsqueeze(0)])

m = A_batch.shape[2]
```

- with a for loop as suggested by @Electric-Turtle 

```
%%timeit
X = torch.stack([torch.lstsq(B_batch[i], A_batch[i])[0][:m] for i in range(n_batch)],dim=0)

# on my machine
# 16.7 ms ± 320 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

- when using batch-wise `torch.pinverse`
```
%%timeit
X = batch_lstsq(B_batch, A_batch)

# on my machine
# 3.02 ms ± 60.7 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**Comment by vishwakftw** [comment]
We have this functionality available in master, so closing. Please reopen if there are any discrepancies.

**Comment by Balandat** [comment]
Note that it is available as `torch.linalg.lstsq`, but `torch.lstsq` does not support batching.
